addons/mail/static/src/components/discuss/tests/discuss_tests.js

```diff
@@ -4150,17 +4150,142 @@ QUnit.test('messages marked as read move to "History" mailbox', async function (
         document.querySelectorAll(`.o_Discuss_thread .o_MessageList_empty`).length,
         0,
-        "History mailbox should not be empty after mark as read"
+        "history mailbox should not be empty after mark as read"
     );
     assert.strictEqual(
         document.querySelectorAll(`.o_Discuss_thread .o_MessageList_message`).length,
         2,
-        "History mailbox should have 2 messages"
+        "history mailbox should have 2 messages"
+    );
+});
+
+QUnit.test('mark a single message as read should only move this message to "History" mailbox', async function (assert) {
+    assert.expect(9);
+
+    this.data['mail.message'].records = [{
+        author_id: [5, 'Demo User'],
+        body: '<p>test 1</p>',
+        id: 1,
+        needaction: true,
+        needaction_partner_ids: [3],
+    }, {
+        author_id: [6, 'Test User'],
+        body: '<p>test 2</p>',
+        id: 2,
+        needaction: true,
+        needaction_partner_ids: [3],
+    }];
+    this.data['mail.notification'].records = [{
+        id: 50,
+        is_read: false,
+        mail_message_id: 1,
+        res_partner_id: 3,
+    }, {
+        id: 51,
+        is_read: false,
+        mail_message_id: 2,
+        res_partner_id: 3,
+    }];
+    this.data.initMessaging.needaction_inbox_counter = 2;
+    await this.start({
+        discuss: {
+            params: {
+                default_active_id: 'mail.box_history',
+            },
+        },
+        env: {
+            session: {
+                name: 'Admin',
+                partner_display_name: 'Your Company, Admin',
+                partner_id: 3,
+                uid: 2,
+            },
+        },
+    });
+    assert.hasClass(
+        document.querySelector(`
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.messaging.history.localId
+            }"]
+        `),
+        'o-active',
+        "history mailbox should initially be the active thread"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_MessageList_empty',
+        "history mailbox should initially be empty"
+    );
+
+    await afterNextRender(() =>
+        document.querySelector(`
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.messaging.inbox.localId
+            }"]
+        `).click()
+    );
+    assert.hasClass(
+        document.querySelector(`
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.messaging.inbox.localId
+            }"]
+        `),
+        'o-active',
+        "inbox mailbox should be active thread after clicking on it"
+    );
+    assert.containsN(
+        document.body,
+        '.o_Message',
+        2,
+        "inbox mailbox should have 2 messages"
+    );
+
+    const movedMessageLocalId = document.querySelectorAll('.o_Message')[0].dataset.messageLocalId;
+    const untouchedMessageLocalId = document.querySelectorAll('.o_Message')[1].dataset.messageLocalId;
```

Comment:
I feel test is more readable when assertions refers to message by their id instead of contextual qualifying adjectives. Not a big deal though.

Reply:
yeah but I don't know how else you could achieve the needed actions in this test ?

Reply:
Tests actions and assertions are just more explicit, like:

- (assertion) should display messages [1, 2, 50, 51] in inbox
- (action) mark messages [1, 2] as read
- (assertion) messages [1, 2] should be in history after being marked as read
- (assertion) messages [50, 51] should still be in inbox (not marked as read)

Reply:
Yeah ok with the idea but I don't know how you can select in DOM the messages you want based on their id ?

Reply:
You get them from local id, and the local id you get it from message id (so that no leak):
```js
document.querySelector(`
   .o_Message[data-message-local-id="${
      this.env.models['mail.message'].find(message => message.id === 51)
   }"]
`);
```
We plan converting `data-*-local-id` to `data-*-id` (+ `data-*-model` with `mail.thread` records), so no use of local id in DOM.

Reply:
As much as I dislike `data-` in the first place, at least having `id` is better than `local-id`.

The `id` is the value you gave it in `data`, you don't need to get it back from the DOM.